google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -1006,9 +1048,10 @@ public void run() {
                 if (closureFuture != null) {
                   closeSession = true;
                 } else {
-                  totalSessions++;
-                  Preconditions.checkState(totalSessions <= options.getMaxSessions());
-                  releaseSession(new PooledSession(session));
+                  Preconditions.checkState(totalSessions() <= options.getMaxSessions() - 1);
+                  PooledSession pooledSession = new PooledSession(session);
+                  allSessions.put(session.getName(), pooledSession);
```

Comment:
Are all session guaranteed to have unique names? It seem that allSessions can and should be a Set, relying on reference-equality of Session objects (PolledSession does not implement equals()/hashCode(), which fits well here).